### DISCUSSION TITLE
Charge and discharge

### DISCUSSION THREAD

**Discussion by kristianolofsson** [wlcrs/huawei_solar#125] [Setting up the integration]
- When ticket the button "elevate permissions" i was not asked to enter the installer pwd. How to know its correct?
eg. I do see the huawei services for charge and discharge.
- I have no battery installed yet, but will have soon. Do i need to reinstall the integration?
- What are the differences in between the different services? Any huawei manual?

Thanks alot for this! My best integration!

**Comment by bengtner** [comment]
I have asked the same question as you about operating modes. I found this page on Huawei support forum:

https://support.huawei.com/enterprise/en/doc/EDOC1100167258/ff9f37a6/setting-the-grid-tied-ess-mode

**Reply by kristianolofsson** [reply]
Sweet! Im running a house-setup with electricity price per hour and is looking for a neat solution. Charge when cheap and discharge when expensive. That's why I would like to know how the different services works. Do you trigger the services from HA or NodeRed?

**Comment by wlcrs** [comment]
* Only needed when connecting via ModbusTCP, I suspect that you use a serial connection instead?
* Restart of HA should be fine to discover them
* See the Wiki: https://github.com/wlcrs/huawei_solar/wiki/Force-charge-discharge-battery

**Reply by kristianolofsson** [reply]
- I use a TCP connection (Fixed) but still no installer pwd required. Can i check somehow its still OK? 
Im an installer when logging in using the portal.

**Reply by wlcrs** [reply]
What do you mean by 'fixed'?

**Reply by kristianolofsson** [reply]
LAN cable

**Reply by wlcrs** [reply]
I suspect that you already have elevated permissions without the need for the installer password via that way. Elevated permissions are only relevant if optimizers and/or a battery is present.

**Reply by kristianolofsson** [reply]
ok, ill reinstall it when battery is in place then to check.
To be clear, i have a dongle to the inverter where i connected a LAN cable, not using Wifi.
The inverter and Dongle got the latest version and only port that seemed to work was the 502.

**Reply by RabAusten** [reply]
I didn't select _Elevated permissions_, is there a way to set this after integration is installed or do I need to remove and reinstall?

**Reply by CommanderROR9** [reply]
I don't think you can activate it after the fact, but as far as I can tell ot doesn't make a difference either way

**Comment by kristianolofsson** [comment]
If i understand the services correct the "forceable charge" set the battery to charge indp. of time of day and Toggle "Charge from grid" should always be on. 
The "forcible discharge" is the opposite, it sets the battery to be used until its drained out. (I guess there is a misspell in the wiki here) - "This allows you to forcibly **charge** your battery"

I would like to use these 2 services to control the feed to my battery depending on the price of my electricity. The price various per hour. - Cheap -> Charge. Expensive -> Discharge

If price is high in between 8-11 AM and 4-6PM id like to discharge. If cheap i charge (normally night-time.

**Reply by bengtner** [reply]
I made a test today with working mode "Charge from grid". I used "Maximum charge/discharge power" start charging/discharging. I think I tried "Forceable charge" first without success. But maybe I had a different working mode by then. Unfortunately I can't find any documentation describing the forceable options.

A straight forward solution is to raise Maximum charge power when price drops below average. And raise discharge when the price is above average. Of course it is possible to find 2 hours with the lowest price and charge by then. And corresponding trick for discharge. But such an automation is a lot more complex to implement. And as long as price drops/increase quickly, the simple approach will work quite well.

**Reply by kristianolofsson** [reply]
So you tried all services but wo any luck? Any errors? Nothing changed at the inverter? Did you use the HA service tab or node red?

**Reply by bengtner** [reply]
I just use plain HA. Made a desktop where I just but all controls/sensors provided by the integration. Just made a new test. "Charge from grid" set true.  Both Maximum chage/discharge set to 0. Then I increased "Forcible charge power" to 2500 and waited for charging to take off. Nothing happened. But if I increase "Maximum charing power" to 2500, charging kicks in. So I can't see any use of  "Forcible charging/discharging power".

**Reply by kristianolofsson** [reply]
ok, so "Charge from grid" controls if grid should be used for charging or only solar? The aim for Max charge, forced or not is to control the limit of power used when charging the batteries? By reading the wiki I can't see any difference in usage of them. Forcing what?

**Reply by bengtner** [reply]
Sorry, I made a mistake in my test. Only way to charge the battery from grid is to use TOU and define a charge segment. Must be done in the FusionSolar app connected to the inverter WLAN. As far as I understand you can control charging within the segments by adjusting the Maximize charge/discharge power parameter. I think you can regard the TOU mode as you introduce a main switch that decides whether you should use solar power or grid for battery charging. In the other modes only solar power can be used for charging.

**Comment by kristianolofsson** [comment]
What I think I need is only to control if the battery should be charged or discharged. Like a Manuel switch. When I do so I want the full possible power.
Because either the price is low, then I want to charge. If it’s high I want to discharge. How to do that then?

**Reply by bengtner** [reply]
Unfortunately it is a bit more complex than that. Se above. Huawei doesn't make this easy.....
You can only charge the battery from grid in charging segments defined under TOU (time of use). The switch "Charge from grid" must be set, but unfortunatly this is not sufficient to make it work. 

If you have defined a charging segment (you can do this statically since lowest price normally occurs during the night). Once you detect the lowprice hours (inside a charging segment), you can start charging by setting Maximum charge power to 2500W. If this value is 0, the battery will not be charged during the night. It might work with the "Charge from grid" switch as well, but I haven't tested this.

In the same way you can control discharging, but then you have to do this in a discharge time segment (when price is high). You start discharge by raising the Maximum discharge power to 2500w. It will automatically stop when the battery is discharged. The "Charge from grid" switch might work here as well in the opposite direction, i.e discharing will start if you turn off this switch. I haven't tested this either. But it works to change the Maximize charging/discharging power.

The other modes (Maximize self consumtion and Fully fed to grid) can automatically charge battery (Maximize self consumtion) or sell as much energy as possible (Fully fed to grid), but it is not possible to control charging using other switches/parameters in those modes.  I have tried.... So if you need detailed control and a possibility to charge the battery from grid, you have to use TOU.

To be honest, I find this scheme pretty awkward but I can't find a simple way to do this. If you come up with another way, please let me know!

I believe Foricible charge/discharge power are associated with battery maintenance procedures. I found those options in the local management system in a tab dealing with battery maintenance. I think it was a mistake to add those in this integration. They can't be used during normal operation as far as I understand. You can't find any info about those parameters on Huawei support forums.

**Reply by kristianolofsson** [reply]
As it seems! Thanks for your deep test in this scenario and great feedback. I will continue with my tests too.
But if only TOU is possible, can the time for charge and discharge be part of the payload to the inverter. Otherwise we are stuck.
Otherwise I was thinking that my logic would be to update the charging and discharging per hour.
<img width="1111" alt="Screenshot 2022-09-01 at 20 49 54" src="https://user-images.githubusercontent.com/7191232/187992814-5dfb5717-b6f3-4e3c-899c-74cd17548f46.png">

**Reply by bengtner** [reply]
I don't really get what you mean by "can the time for charge and discharge be part of the payload to the inverter."? The time segments used in TOU mode is fixed (or pseudo-fixed - it is not that easy to update those). Let's say you decide to define a charging segment between 9pm and 6am and a discharge segment the rest of the day.  If you can identify the cheapest hours during night, you can just set  Maximize charging power to 2500W during those hours and the battery will be charged by grid. If you discover a few cheap hours at noon and you set Maximize charging power by then, nothing will happen since this falls in the discharge segment.  To trigger discharging you do a similar operation. Identify the most expensive hours and set Maximize discharge power by then. This will result in discharge of the battery when the price peaks. 

So you need to define the segments based on historical data to maximize the probability the extremes falls within the the right segment.

**Reply by kristianolofsson** [reply]
What I meant is that if time period can be part of the payload I can set the charge and discharge from HA. eg for above. 
00:00 - 04:00 Charge, 04:00-19:00 Discharge
I also found this:
When this parameter is set to Time-of-Use Price, discharge is performed at a high price, and the charge is performed at a low price. A maximum of **10-time segments** can be added. When Time-of-Use Electricity Price is set, the number of time segments is greater than or equal to 2 and the number of electricity prices is greater than or equal to 2.
Setting grid charging to "enable" allows the grid to charge energy storage devices. The factory default is "disable" and users who enable this function must comply with local regulations for grid charging.

**Reply by bengtner** [reply]
According to my opinion it is a lot easier to keep the schedule in HA and change the operating mode as I describe below. Today there is no support in this integration to adjust the schedule and it is a significant amount of work to add support for this. Apply my model and there is no need for this support.

**Comment by bengtner** [comment]
Battery control is extremely complicated. Huawei system behaviour is not easy to understand partly caused by limited documentation, but also by a strange design according to me. After a few more tests I have to revise my earlier writing (again).

Tonight I discovered battery discharge under TOU mode, will only be made at rate that tops up solar production **to cover your own consumtion**. So TOU can't be used if you plan to sell battery energi to grid.

So, if you would like to implement a "buy at low-price and sell at high-price" strategy (BLSH), I believe this is the best approach:

a) Define one charge segment covering almost all hours in the local management system. You need to have one discharge segment as well, but you can make this very short. You will control discharge with a different operational mode than TOU. Select this short discharge segment at a time where it is unlikely you would like to charge the battery, e.g in the morning hours where the electricity price tends to peak. EDIT: It seems like this segment can't be shorter than 1h. Otherwise charging will fail. See a separate post about this.
b) Make sure Maximum charge power is set to comply to your battery capacaity. 2500W for a standard battery. Also activate Charge from grid. Set Maximum discharge power to  "max inverter capacity" - "max solar panel capacity". This will ensure the sum of battery discharging and  solar production will not exceed your inverter capacity. The reason why I propose this limit is I don't know what source will be prioritized if you decide to sell battery energy when solar production peaks and the sum of those two exceeds the inverter capacity.
c) Select TOU when you would like to charge the battery. Use Maximize self-consumtion during normal operation. Since the battery normally is fully loaded, this operational mode will not "steal" any solar production to load the battery. Excess solar energy will be fed to the grid automatically as long as the battery  is fully loaded.
d) Select Fully fed to grid when you want to sell battery energy. **This is the only mode where the battery can be fed to the grid!**

If yoy want to implement a BLSH strategy, you need to implement the proper automations to change the operating mode. Preferably this should be done with the electricity price as input. This will require quite some extra work, and I will not cover this in this post.

I you can't implement automatic control of the operational mode and don't want to manually update this 2 or 3 times a day, I suggest the following setup:

a) Go for TOU mode and define a charging segement for those hours the electricy price normally is low during the night. Define a discharge segment in the evening when the electricy price peaks and solar production is low. By this (cheap) battery energy will be used to top up solar production just before the energy price drops at night. Then the battery will be charged again during the night.
b) Avoid to define a discharge segment in the morning. Doing so will drain the battery, and solar energy will be used to charge the battery during the day (unless disabling the Charge from grid option will prevent this - I'm not sure if this is the case). It is better to sell solar energy than to load the battery during day time.

**Reply by kristianolofsson** [reply]
Ok, so you where thinking that we keep the time segments outside of Huawei. 
a)TOU - 00:00-24:00 (Charge) and then 08:00 - 08:01 (Discharge). This well allow us to charge at all time.
b) Set the parameters once and "never" change them.
c) Time will control the mode by switching from TOU and FFTG mode. (Charge/Discharge)

My plan is to keep the timing and algorithm in HA. HA will decide if to Charge or discharge like my graph above.
The graph and result from graph calculates if I should charge or discharge depending Battery size, future price trend etc.

Green is discharge and red is charge

**Reply by bengtner** [reply]
Exactly what I mean. This makes control quite easy and understandable. Just think of TOU as charging from grid, "Fully fed to grid" as discharge and the last mode as normal operation.

**Reply by JohanEkblad** [reply]
Great to know! I will try this when i get my inverter and batteries. I am going to integrate this to nordpool

**Reply by MichaelPihlblad** [reply]
Thanks. Same here. Will try as soon as the battery is installed.

Best approach to utilize script function i guess.  Would be great to share and collaborate on the algorithm.

**Reply by bengtner** [reply]
The reason behind my approach is it is quite tricky to implement algorithms in HA scripts. Python is a lot easier. I might move my python script into HA as an integration, but that will require some modifications of the code (HA requires an asynchronous execution model)

**Reply by wlcrs** [reply]
Have you looked into using AppDaemon as the environment for your Python script? That way you can easily call the HA-services that this integration exposes (or change the `number` entities to limit charging/discharging).

An easy hook can be [run_minutely](https://appdaemon.readthedocs.io/en/latest/AD_API_REFERENCE.html#appdaemon.adapi.ADAPI.run_minutely) or `run_hourly` to compute if you need to change those parameters.

**Reply by bengtner** [reply]
My script is already running and interacts with HA over the REST interface, setting the battery working mode to charge/discharge/idle. Works fine.

**Reply by kristianolofsson** [reply]
Thanks again for your valuable input. So to rap up.
I will define the TOU and charging parameters according to your suggestion above. I will then use an automatic price algorithm to decide if the mode price is “Cheap”, “Normal” and “expensive”. The corresponding modes will be TOU, MSC and FFTG.
Bear in mind that the algorithm will relay to price over time and battery capacity. So “expensive” is in relation to price over time.

**Reply by bengtner** [reply]
Today I noticed the battery did not charge from grid. I think this problem appeared when I reduced the discharge segment in the TOU setting down to 10min. I just use TOU for charging, so I wanted to minimize the time period when I'm not allowed to charge. When I restored the discharge segment to 1h, charging started to work again. I don't know it there is any limitation causing this issue, or just a bug that happened once when I made a change of the segments.

**Comment by kristianolofsson** [comment]
I got the algorithm set and is evaluating. The result will return charge/discharge. I know need some help with the modbus payload to set the same. Anyone got something that works?

**Reply by wlcrs** [reply]
Maybe try using AppDaemon to integrate your algorithm into HA? Otherwise it will not be possible to use the integration concurrently: only one Modbus connection at a time is supported by the inverter

**Reply by bengtner** [reply]
See my earlier post(s). If you have a standalone script, it's easy to use the HA REST interface trigger the HA service that changes the battery working mode. And then you are done!

**Reply by kristianolofsson** [reply]
I have now made some more tests and its not that easy as it should be.
I define the inverter in Maximum self consumtion and then only use the Forced charging function.
It seems to work pretty well. The forced charging service overides the std mode and starts charging from grid.
But the service do seomtimes end up in a service call error. Its seems to be timeout, waiting the the response. How to solve this?

```
Logger: homeassistant.components.websocket_api.http.connection
Source: custom_components/huawei_solar/services.py:168
Integration: Home Assistant WebSocket API (documentation, issues)
First occurred: 14:50:03 (4 occurrences)
Last logged: 17:00:04

[140497763785152]
[140497909099264]
[140498897339728]
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 456, in wait_for
    return fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 200, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1738, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1775, in _execute_service
    await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
  File "/config/custom_components/huawei_solar/services.py", line 168, in stop_forcible_charge
    await bridge.set(rn.STORAGE_FORCIBLE_DISCHARGE_POWER, 0)
  File "/usr/local/lib/python3.10/site-packages/huawei_solar/bridge.py", line 359, in set
    return await self.client.set(name, value, slave=self.slave_id)
  File "/usr/local/lib/python3.10/site-packages/huawei_solar/huawei_solar.py", line 496, in set
    response = await self.write_registers(
  File "/usr/local/lib/python3.10/site-packages/huawei_solar/huawei_solar.py", line 520, in write_registers
    response = await self._client.protocol.write_registers(
  File "/usr/local/lib/python3.10/site-packages/pymodbus/client/asynchronous/async_io/__init__.py", line 35, in execute
    resp = await asyncio.wait_for(req, timeout=self._timeout)
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 458, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError
```

**Reply by bengtner** [reply]
I guess you have a problem with the connection that creates this fault. What interface do you use to connect to Modbus? 

I'm running Modbus over TCP using the sDongle. and this is stable.

**Reply by wlcrs** [reply]
This error indicates that the inverter did not respond within 5 seconds (the default timeout) to the command sent by the integration.

In the case of read-commands, I just retry the command (up to 10 times) in case of a timeout. However, for write-commands, I do not do this because it is a more sensitive operation. I prefer to bubble up the problem to the end-user instead, as he can determine better what to do. (retry, cancel, ...)

**Comment by kristianolofsson** [comment]
Me too. I’m using this integration and connects over TCP. I get all the values and they frequently updated. This problem is only related to the call service. I find the IP on my router so I presume it’s correct. I use nodered and the home assistant websocket connection. I’ll try the automation in HA to check if it’s better.

**Reply by bengtner** [reply]
OK. That explains the websocket messages. I have just used the this integration in HA. It works fine. The only thing I can notice in the log is  a warning that it takes more than 10s to setup the Huawei integration. But this seems just to be a notice.

**Comment by kristianolofsson** [comment]
I’ll set up an automation in HA instead to check. It seems to be time out error right? No way to extend the time used?

**Comment by kristianolofsson** [comment]
Understood! It’s happening occasionally but still enjoying when it fails. For example It seems like the battery mode switch is more stabile since I guess then the service call to change the HA switch is not reallying on the integration to the modbus. I need to test more to see if the actual mode-change is failing to Huawei there too.

Would a serial connection be more stabile?

Kristian

Skickat från Outlook för iOS<https://aka.ms/o0ukef>
________________________________
Från: Thijs W. ***@***.***>
Skickat: Tuesday, September 27, 2022 1:49:15 PM
Till: wlcrs/huawei_solar ***@***.***>
Kopia: Kristian O ***@***.***>; Author ***@***.***>
Ämne: Re: [wlcrs/huawei_solar] Charge and discharge (Discussion #125)


This error indicates that the inverter did not respond within 5 seconds (the default timeout) to the command sent by the integration.

In the case of read-commands, I just retry the command (up to 10 times) in case of a timeout. However, for write-commands, I do not do this because it is a more sensitive operation. I prefer to bubble up the problem to the end-user instead, as he can determine better what to do. (retry, cancel, ...)

**Reply by bengtner** [reply]
Beside the warning at startup (integrations takes more than 10s to initiate), I occasionally see  this warning in my log:
```
Logger: homeassistant.helpers.service
Source: helpers/service.py:635
First occurred: 01:01:09 (2 occurrences)
Last logged: 06:01:22

Unable to find referenced entities select.battery_working_mode or it is/they are currently not availab
```
So this relates to the battery working mode. On request from my Python script, HA will update (using this integration)  this setting to either "TOU", "Fully Fed to Grid"
or "Maximise self consumption". This night this warning happened twice. But it seems the operation towards the battery was carried out anyway.  I have made a few restarts of HA due to config changes recently, so my log is not that long. I will keep an eye on this to see if I have problems with any other entities.

**Reply by MilletGtR** [reply]
Hello bengtner, would you consider sharing your setup with scripts and all? I've been following this thread closely and am very interested in how you managed to solve the battery mode.

**Reply by bengtner** [reply]
Yes, I can do that. Give me a day or two, and i will publish my script in a separate repo on GitHub. I need to wash out some stuff first ...

**Reply by MilletGtR** [reply]
Thank you so much, that would help us tremendously.

**Reply by bengtner** [reply]
Please find my code and som documenatation in the README.md file.

https://github.com/bengtner/batterycontrol.git

Enjoy!

**Reply by MilletGtR** [reply]
Such a good and detailed explanation, this is very advanced stuff. Thank you for taking the time bengtner!

**Reply by kristianolofsson** [reply]
> Please find my code and som documenatation in the README.md file.
> 
> https://github.com/bengtner/batterycontrol.git
> 
> Enjoy!

Can the script be implemented as part of HA?
Can the script be triggered manually and by then get an update of the help entity?
Can the total array be pushed as attributes to the help entity so the future can be plotted with eg a graph

**Reply by bengtner** [reply]
Sorry for the late answer.

No, the script cannot be executed as a part of HA without major changes.
No, the script is written for continous operation. It will trigger HA every hour.
No, see my earlier answer

**Comment by steinaaj** [comment]
Hi Kristian - I'm also trying to find a solution to maximise the battery use. However I can't wrap my mind around .yaml

Have you tried Node Red? I can't seem to find the correct "device_id" for the battery...

**Reply by kristianolofsson** [reply]
Yes I use nodered for my control. I use an algorithm similar to above and then depending on the result I set a helper input node that controls the inverter.

**Reply by steinaaj** [reply]
Could you take a screenshot of the nodered automation? Which device_id are you using for the battery?

**Reply by nixtagangsta** [reply]
Hi
is it possible to get a more detailed explanition how to set up the battery settings in node-red? i have created a flow like the picture but cant figure out the settings.

Niklas

**Reply by nixtagangsta** [reply]
> Hi is it possible to get a more detailed explanition how to set up the battery settings in node-red? i have created a flow like the picture but cant figure out the settings.
> 
> Niklas

Anybody?

**Reply by nixtagangsta** [reply]
> > Hi is it possible to get a more detailed explanition how to set up the battery settings in node-red? i have created a flow like the picture but cant figure out the settings.
> > Niklas
> 
> Anybody?

Nobody? Could someone export their node red config and post it here please?

**Comment by kristianolofsson** [comment]
<img width="1194" alt="Screenshot 2022-11-21 at 17 18 30" src="https://user-images.githubusercontent.com/7191232/203105454-8a1210ef-a707-43ee-b52c-8485c36a52e1.png">

**Reply by steinaaj** [reply]
--- edit ---
Nevermind - found the ID in the dev-tools :D 
Thx again!
---

That is amazing - thx a bunch!
One last question; which device_id (battery) are you calling? I can't seem to figure out where to find the name / id?

**Comment by CommanderROR9** [comment]
Ok...I tried to read through most of the replies ...
Am I correct in the assumption that there is currently no way to directly trigger an AC Charge from Home Assistant in an automation, outside the workaround with the TOU mode set up with 23h a day set as "low cost"?

**Reply by steinaaj** [reply]
You can send "forcible_charge" and "stop_forcible_charge". It appears to be working in every mode.

**Reply by CommanderROR9** [reply]
Ok... but I don't seem to have any entities for that? I will have to look into it, it might be because I set up the integration before I added the Luna2000 to my setup...

**Reply by CommanderROR9** [reply]
Ok...I deleted and reinstalled the integration but still don't have the "force charge" options. I guess I will have to do some more research on that then.

**Reply by MichaelPihlblad** [reply]
It's  a service not a state. You can browse it in developer menu under services

**Reply by CommanderROR9** [reply]
Thanks, I have found it

**Comment by kristianolofsson** [comment]
Check the services in the dev page

**Comment by CommanderROR9** [comment]
Thanks. I had just found that. Playing with it now...but running into some
snags with timeouts

Kristian O ***@***.***> schrieb am So., 27. Nov. 2022, 23:07:

**Comment by CommanderROR9** [comment]
Ok, managed to get it working. Thanks for your help!

**Comment by Ulmerro** [comment]
I have a question, if i make force charge i have no values in the fusion solar app  Pv is 0 what is this for a problem.
